code/__HELPERS/mobs.dm

```diff
@@ -346,62 +346,92 @@ This is always put in the attack log.
 	if(progress)
 		qdel(progbar)
 
-/proc/do_after(mob/user, delay, needhand = 1, atom/target = null, progress = 1, datum/callback/extra_checks = null)
+/*	Use this proc when you want to have code under it execute after a delay, and ensure certain conditions are met during that delay...
+ *	Such as the user not being interrupted via getting stunned or by moving off the tile they're currently on.
+ *
+ *	Example usage:
+ *
+ *	if(do_after(user, 50, target = sometarget, extra_checks = list(callback_check1, callback_check2)))
+ *		do_stuff()
+ *	
+ *	This will create progress bar that lasts for 5 seconds. If the user doesn't move or otherwise do something that would cause the checks to fail in those 5 seconds, do_stuff() would execute.
+ *	The Proc returns TRUE upon success (the progress bar reached the end), or FALSE upon failure (the user moved or some other check failed)
+ */
+/proc/do_after(mob/user, delay, needhand = 1, atom/target = null, progress = 1, list/extra_checks = list(), use_default_checks = TRUE)
 	if(!user)
-		return 0
+		return FALSE
 	var/atom/Tloc = null
 	if(target)
 		Tloc = target.loc
 
 	var/atom/Uloc = user.loc
 
-	var/drifting = 0
+	var/drifting = FALSE
 	if(!user.Process_Spacemove(0) && user.inertia_dir)
-		drifting = 1
+		drifting = TRUE
 
 	var/holding = user.get_active_hand()
 
-	var/holdingnull = 1 //User's hand started out empty, check for an empty hand
+	var/holdingnull = TRUE //User's hand started out empty, check for an empty hand
 	if(holding)
-		holdingnull = 0 //Users hand started holding something, check to see if it's still holding that
+		holdingnull = FALSE //Users hand started holding something, check to see if it's still holding that
 
 	var/datum/progressbar/progbar
 	if(progress)
 		progbar = new(user, delay, target)
 
 	var/endtime = world.time + delay
 	var/starttime = world.time
-	. = 1
+	. = TRUE
+
+	// By default, checks for weakness and stunned get added to the extra_checks list.
+	// Setting `use_default_checks` to FALSE means that you don't want the do_after to check for these statuses, or that you will be supplying your own checks.
+	if(use_default_checks)
+		extra_checks += CALLBACK(user, /mob.proc/IsWeakened)
+		extra_checks += CALLBACK(user, /mob.proc/IsStunned)

```

Comment:
Careful - `extra_checks` will have these callbacks added on every call of the function, and since `extra_checks` is a singleton (modifications to it it will be retained across function calls) it'll have more and more copies of these callbacks added for every call that uses `use_default_checks`.

Instead, you may want to have `extra_checks` be `null`, and then if not supplied, replace it with a list that contains these two callbacks.

Reply:
I'm not sure what you mean here, every time something calls do_after, `extra_checks` starts as a blank list, provided the coder didn't supply anything for the argument. But even then, it should be unique per proc call.

I just tested this out in game. I added some print statements for the amount of callbacks in `extra_checks`. Then I spammed clicked a thing which created uses a do_after a bunch of times. (tested with limb regrowing, but also cutting wires out of a SMES). The amount of callbacks in `extra_checks` was always 2.

```js
for(var/datum/callback/C in extra_checks)
    message_admins("[C]")
```

![image](https://user-images.githubusercontent.com/42044220/75634351-b7046600-5bd2-11ea-9d28-9cd92be80747.png)

Reply:
Oh - it looks like this isn't the case in BYOND, good. I think I might have had a wire crossed from python, where that pattern would cause the list to balloon in size